scripts/Packages/Package_BGA/ipc_bga_generator.py

```diff
@@ -18,8 +18,38 @@
 from string import ascii_uppercase
 
 def generateFootprint(config, fpParams, fpId):
-    print('Building footprint for parameter set: {}'.format(fpId))
+    createFp = False
     
+    # use IPC-derived pad size if possible, then fall back to user-defined pads
+    if "ball_type" in fpParams and "ball_diameter" in fpParams:
+        try:
+            padSize = configuration[fpParams["ball_type"]]
+            try:
+                padSize = configuration[fpParams["ball_type"]][fpParams["ball_diameter"]]["max"]
+                fpParams["pad_size"] = [padSize, padSize]
+                createFp = True
+            except KeyError as e:
+                print("{}mm is an invalid ball diameter. See ipc_7351b_bga_land_patterns.yaml for valid values. No footprint generated.".format(e))
+        except KeyError:
+            print("{} is an invalid ball type. See ipc_7351b_bga_land_patterns.yaml for valid values. No footprint generated.".format(e))
+        
+        if "pad_diameter" in fpParams:
+            print("Pad size is being derived using IPC rules even though pad diameter is defined.")
+    elif "ball_type" in fpParams and not "ball_diameter" in fpParams:
+        raise KeyError("Ball diameter is missing. No footprint generated.")
+    elif "ball_diameter" in fpParams and not "ball_type" in fpParams:
+        raise KeyError("Ball type is missing. No footprint generated.")
+    elif "pad_diameter" in fpParams:
+        fpParams["pad_size"] = [fpParams["pad_diameter"], fpParams["pad_diameter"]]
+        print("Pads size is set by the footprint definition. This should only be done for manufacturer-specific footprints.")
+        createFp = True
+    else:
+        raise KeyError("The config file must include 'ball_type' and 'ball_diameter' or 'pad_diameter'. No footprint generated.")
```

Comment:
I don't like that the behavior is not consistent. Some branches throw a KeyError, others print a message (and keep `createFp = false`)
I would change them all to just print the error message.
Or is there a good reason for the current scheme that I am missing?

Reply:
OK Changed them all to print(). I think this is rather inconsistent in all the IPC generator scripts (and maybe others, too).